File: k8s/testing/load-testing/k6-scripts/epic8-load-test-suite.js
/**
 * Epic 8 Comprehensive Load Testing Suite
 * K6 script for testing auto-scaling and performance under load
 * Swiss engineering precision with detailed metrics collection
 */

import http from 'k6/http';
import { check, sleep } from 'k6';
import { Rate, Trend, Counter, Gauge } from 'k6/metrics';
import { randomIntBetween, randomItem } from 'https://jslib.k6.io/k6-utils/1.2.0/index.js';

// Epic 8 Custom Metrics
const epic8Latency = new Trend('epic8_end_to_end_latency', true);
const epic8ErrorRate = new Rate('epic8_error_rate');
const epic8Throughput = new Counter('epic8_requests_total');
const epic8ConcurrentUsers = new Gauge('epic8_concurrent_users');
const epic8CostPerQuery = new Trend('epic8_cost_per_query_usd');
const epic8SuccessRate = new Rate('epic8_success_rate');

// Service-specific metrics
const apiGatewayLatency = new Trend('epic8_api_gateway_latency');
const generatorLatency = new Trend('epic8_generator_latency');
const analyzerLatency = new Trend('epic8_analyzer_latency');
const retrieverLatency = new Trend('epic8_retriever_latency');
const cacheHitRate = new Rate('epic8_cache_hit_rate');

// Configuration
const BASE_URL = __ENV.EPIC8_BASE_URL || 'https://api.epic8.platform';
const API_VERSION = __ENV.API_VERSION || 'v1';
const API_KEY = __ENV.EPIC8_API_KEY || 'test-api-key';

// Test data - Realistic RAG queries
const testQueries = [
  "What are the best practices for Kubernetes auto-scaling?",
  "How do I optimize vector search performance in production?",
  "Explain the differences between HPA and VPA in Kubernetes",
  "What are the cost optimization strategies for multi-cloud deployments?",
  "How does model routing work in multi-model RAG systems?",
  "What are the Swiss engineering principles for system reliability?",
  "How to implement circuit breaker patterns in microservices?",
  "What metrics should I monitor for ML model performance?",
  "How to handle cache invalidation in distributed systems?",
  "What are the best practices for prompt engineering in RAG?",
  "How to implement proper observability in cloud-native applications?",
  "What are the security considerations for RAG systems?",
  "How to optimize database queries for vector similarity search?",
  "What are the trade-offs between accuracy and latency in ML systems?",
  "How to implement blue-green deployments with zero downtime?"
];

// Load test scenarios configuration
export let options = {
  scenarios: {
    // Baseline load test - Normal traffic pattern
    baseline_load: {
      executor: 'ramping-vus',
      startVUs: 10,
      stages: [
        { duration: '2m', target: 50 },   // Ramp up to 50 users
        { duration: '5m', target: 50 },   // Stay at 50 users
        { duration: '2m', target: 100 },  // Ramp up to 100 users
        { duration: '10m', target: 100 }, // Stay at 100 users
        { duration: '2m', target: 0 },    // Ramp down
      ],
    },

    // Peak load test - Traffic spikes
    peak_load: {
      executor: 'ramping-vus',
      startVUs: 50,
      stages: [
        { duration: '1m', target: 200 },  // Quick ramp to 200 users
        { duration: '3m', target: 500 },  // Spike to 500 users
        { duration: '5m', target: 500 },  // Sustain peak load
        { duration: '2m', target: 200 },  // Gradual decrease
        { duration: '1m', target: 0 },    // Ramp down
      ],
      startTime: '15m', // Start after baseline
    },

    // Stress test - Beyond normal capacity
    stress_test: {
      executor: 'ramping-vus',
      startVUs: 100,
      stages: [
        { duration: '2m', target: 500 },  // Ramp to 500 users
        { duration: '3m', target: 1000 }, // Stress to 1000 users
        { duration: '5m', target: 1000 }, // Sustain stress
        { duration: '2m', target: 500 },  // Step down
        { duration: '2m', target: 0 },    // Ramp down
      ],
      startTime: '25m', // Start after peak load
    },

    // Spike test - Sudden load increase
    spike_test: {
      executor: 'ramping-vus',
      startVUs: 100,
      stages: [
        { duration: '30s', target: 100 }, // Normal load
        { duration: '30s', target: 1500 }, // Sudden spike
        { duration: '1m', target: 1500 },  // Sustain spike
        { duration: '30s', target: 100 },  // Return to normal
        { duration: '30s', target: 0 },    // Ramp down
      ],
      startTime: '40m', // Start after stress test
    },

    // Endurance test - Long duration
    endurance_test: {
      executor: 'constant-vus',
      vus: 200,
      duration: '30m',
      startTime: '45m', // Start after spike test
    },

    // Auto-scaling validation test
    autoscaling_test: {
      executor: 'ramping-arrival-rate',
      startRate: 10,
      timeUnit: '1s',
      stages: [
        { duration: '2m', target: 20 },   // 20 RPS
        { duration: '2m', target: 50 },   // 50 RPS
        { duration: '2m', target: 100 },  // 100 RPS
        { duration: '2m', target: 200 },  // 200 RPS - Should trigger scaling
        { duration: '5m', target: 200 },  // Sustain to validate scaling
        { duration: '2m', target: 50 },   // Scale down
        { duration: '2m', target: 10 },   // Return to baseline
      ],
      startTime: '80m', // Start after endurance
    }
  },

  // Global thresholds for Swiss engineering precision
  thresholds: {
    // System-wide SLOs
    'epic8_end_to_end_latency': [
      'p(95)<2000',     // 95% under 2s (SLO requirement)
      'p(99)<5000',     // 99% under 5s
    ],
    'epic8_error_rate': ['rate<0.01'],      // Error rate < 1%
    'epic8_success_rate': ['rate>0.995'],   // Success rate > 99.5%

    // Service-specific thresholds
    'epic8_api_gateway_latency': ['p(95)<500'],  // API Gateway < 500ms
    'epic8_generator_latency': ['p(95)<1500'],   // Generator < 1.5s
    'epic8_analyzer_latency': ['p(95)<200'],     // Analyzer < 200ms
    'epic8_retriever_latency': ['p(95)<100'],    // Retriever < 100ms
    'epic8_cache_hit_rate': ['rate>0.90'],       // Cache hit rate > 90%

    // Cost and efficiency thresholds
    'epic8_cost_per_query_usd': ['p(95)<0.01'], // Cost < $0.01 per query

    // HTTP-level thresholds
    'http_req_duration': ['p(95)<2000'],
    'http_req_failed': ['rate<0.01'],
  },
};

// Setup function - runs once before test
export function setup() {
  console.log('🚀 Starting Epic 8 Load Testing Suite');
  console.log(`📍 Target: ${BASE_URL}`);
  console.log(`🔑 API Version: ${API_VERSION}`);

  // Validate API connectivity
  const healthCheck = http.get(`${BASE_URL}/health`);
  check(healthCheck, {
    'API is healthy': (r) => r.status === 200,
  });

  return {
    baseUrl: BASE_URL,
    apiVersion: API_VERSION,
    timestamp: new Date().toISOString(),
  };
}

// Main test function
export default function(data) {
  const startTime = Date.now();

  // Update concurrent users metric
  epic8ConcurrentUsers.add(__VU);

  // Select random query for realistic testing
  const query = randomItem(testQueries);
  const complexity = getQueryComplexity(query);

  // Construct request
  const payload = JSON.stringify({
    query: query,
    options: {
      max_tokens: 500,
      temperature: 0.7,
      include_sources: true,
      model_preference: randomItem(['auto', 'fast', 'balanced', 'quality']),
    },
    metadata: {
      user_id: `load_test_user_${__VU}`,
      session_id: `session_${__ITER}`,
      complexity: complexity,
      timestamp: new Date().toISOString(),
    }
  });

  const params = {
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${API_KEY}`,
      'X-Epic8-Test': 'load-testing',
      'X-Epic8-VU': __VU.toString(),
      'X-Epic8-Iteration': __ITER.toString(),
    },
    timeout: '30s',
  };

  // Execute end-to-end RAG query
  const response = http.post(`${BASE_URL}/${API_VERSION}/query`, payload, params);

  const endTime = Date.now();
  const duration = endTime - startTime;

  // Record metrics
  epic8Latency.add(duration);
  epic8Throughput.add(1);

  // Validate response
  const success = check(response, {
    'Status is 200': (r) => r.status === 200,
    'Response has answer': (r) => {
      try {
        const body = JSON.parse(r.body);
        return body.answer && body.answer.length > 0;
      } catch {
        return false;
      }
    },
    'Response has sources': (r) => {
      try {
        const body = JSON.parse(r.body);
        return body.sources && Array.isArray(body.sources);
      } catch {
        return false;
      }
    },
    'Response time acceptable': (r) => r.timings.duration < 5000,
    'No server errors': (r) => r.status < 500,
  });

  // Record success/error metrics
  epic8SuccessRate.add(success);
  epic8ErrorRate.add(!success);

  // Parse response for detailed metrics
  if (response.status === 200) {
    try {
      const body = JSON.parse(response.body);

      // Service-specific latencies (if provided in response)
      if (body.metrics) {
        if (body.metrics.api_gateway_latency) {
          apiGatewayLatency.add(body.metrics.api_gateway_latency);
        }
        if (body.metrics.generator_latency) {
          generatorLatency.add(body.metrics.generator_latency);
        }
        if (body.metrics.analyzer_latency) {
          analyzerLatency.add(body.metrics.analyzer_latency);
        }
        if (body.metrics.retriever_latency) {
          retrieverLatency.add(body.metrics.retriever_latency);
        }
        if (body.metrics.cache_hit !== undefined) {
          cacheHitRate.add(body.metrics.cache_hit ? 1 : 0);
        }
        if (body.metrics.cost_usd) {
          epic8CostPerQuery.add(body.metrics.cost_usd);
        }
      }
    } catch (e) {
      console.warn(`Failed to parse response metrics: ${e.message}`);
    }
  } else {
    console.warn(`Request failed: ${response.status} - ${response.body}`);
  }

  // Realistic user behavior - think time
  const thinkTime = getThinkTime(complexity);
  sleep(thinkTime);
}

// Utility functions
function getQueryComplexity(query) {
  const length = query.length;
  const keywords = ['best practices', 'optimization', 'performance', 'architecture'];
  const hasComplexKeywords = keywords.some(keyword => query.toLowerCase().includes(keyword));

  if (length > 80 || hasComplexKeywords) {
    return 'complex';
  } else if (length > 40) {
    return 'medium';
  } else {
    return 'simple';
  }
}

function getThinkTime(complexity) {
  switch (complexity) {
    case 'simple':
      return randomIntBetween(1, 3);  // 1-3 seconds
    case 'medium':
      return randomIntBetween(2, 5);  // 2-5 seconds
    case 'complex':
      return randomIntBetween(3, 8);  // 3-8 seconds
    default:
      return randomIntBetween(2, 4);  // 2-4 seconds
  }
}

// Teardown function - runs once after test
export function teardown(data) {
  console.log('🏁 Epic 8 Load Testing Complete');
  console.log(`📊 Test started at: ${data.timestamp}`);
  console.log(`🎯 Target: ${data.baseUrl}`);

  // Additional cleanup or reporting could go here
}

// Health check scenario for continuous monitoring
export function healthCheck() {
  const response = http.get(`${BASE_URL}/health`);
  check(response, {
    'Health check successful': (r) => r.status === 200,
    'Health check fast': (r) => r.timings.duration < 1000,
  });
}

// Auto-scaling trigger scenario
export function autoScalingTrigger() {
  // Simulate load pattern that should trigger HPA scaling
  const burst = 10; // Number of concurrent requests
  const promises = [];

  for (let i = 0; i < burst; i++) {
    const query = randomItem(testQueries);
    const payload = JSON.stringify({
      query: query,
      options: { max_tokens: 500 },
      metadata: { trigger_test: true }
    });

    promises.push(
      http.asyncRequest('POST', `${BASE_URL}/${API_VERSION}/query`, payload, {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${API_KEY}`,
          'X-Epic8-Test': 'autoscaling-trigger',
        }
      })
    );
  }

  // Wait for all requests to complete
  const responses = Promise.all(promises);

  check(responses, {
    'Burst requests successful': (rs) => rs.every(r => r.status === 200),
  });
}

File: archive/epic8-services/k8s/testing/load-testing/gatling-tests/Epic8LoadTestSuite.scala
package epic8.platform

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import scala.concurrent.duration._
import scala.util.Random

/**
 * Epic 8 Comprehensive Load Testing Suite - Gatling Implementation
 * Swiss engineering precision with advanced auto-scaling validation
 *
 * Test Scenarios:
 * 1. Baseline Load (100 users)
 * 2. Peak Load (500 users)
 * 3. Stress Test (1000 users)
 * 4. Spike Test (1500 users burst)
 * 5. Endurance Test (200 users for 30 minutes)
 * 6. Auto-scaling Validation
 */
class Epic8LoadTestSuite extends Simulation {

  // Configuration
  val baseUrl = System.getProperty("epic8.baseUrl", "https://api.epic8.platform")
  val apiVersion = System.getProperty("epic8.apiVersion", "v1")
  val apiKey = System.getProperty("epic8.apiKey", "test-api-key")
  val environment = System.getProperty("epic8.environment", "staging")

  // HTTP Protocol Configuration
  val httpProtocol = http
    .baseUrl(baseUrl)
    .acceptHeader("application/json")
    .contentTypeHeader("application/json")
    .authorizationHeader(s"Bearer $apiKey")
    .userAgentHeader("Epic8-Gatling-LoadTest/1.0")
    .header("X-Epic8-Test", "gatling-load-testing")
    .header("X-Epic8-Environment", environment)
    // Swiss engineering precision - detailed connection settings
    .connectionHeader("keep-alive")
    .maxConnectionsPerHost(10)
    .shareConnections

  // Test Data - Realistic RAG queries with complexity classification
  val queries = Array(
    ("What are the best practices for Kubernetes auto-scaling?", "complex"),
    ("How do I optimize vector search performance in production?", "complex"),
    ("Explain the differences between HPA and VPA in Kubernetes", "medium"),
    ("What are the cost optimization strategies for multi-cloud deployments?", "complex"),
    ("How does model routing work in multi-model RAG systems?", "complex"),
    ("What are the Swiss engineering principles for system reliability?", "medium"),
    ("How to implement circuit breaker patterns in microservices?", "medium"),
    ("What metrics should I monitor for ML model performance?", "complex"),
    ("How to handle cache invalidation in distributed systems?", "medium"),
    ("What are the best practices for prompt engineering in RAG?", "medium"),
    ("How to implement proper observability in cloud-native applications?", "complex"),
    ("What are the security considerations for RAG systems?", "complex"),
    ("How to optimize database queries for vector similarity search?", "complex"),
    ("What are the trade-offs between accuracy and latency in ML systems?", "complex"),
    ("How to implement blue-green deployments with zero downtime?", "medium"),
    ("What is Kubernetes?", "simple"),
    ("How to scale pods?", "simple"),
    ("What is a RAG system?", "simple"),
    ("How to monitor applications?", "simple"),
    ("What is auto-scaling?", "simple")
  )

  // Model preferences for realistic testing
  val modelPreferences = Array("auto", "fast", "balanced", "quality")

  // Feeders for test data
  val queryFeeder = queries.circular
  val modelFeeder = modelPreferences.circular
  val userIdFeeder = Iterator.continually(Map("userId" -> s"gatling_user_${Random.nextInt(10000)}"))

  // Scenarios

  // Core RAG Query Scenario
  val ragQueryScenario = scenario("RAG Query")
    .feed(queryFeeder)
    .feed(modelFeeder)
    .feed(userIdFeeder)
    .exec(
      http("Execute RAG Query")
        .post(s"/$apiVersion/query")
        .body(StringBody(session => {
          val (query, complexity) = session("queries").as[(String, String)]
          val modelPref = session("modelPreferences").as[String]
          val userId = session("userId").as[String]

          s"""
          {
            "query": "$query",
            "options": {
              "max_tokens": 500,
              "temperature": 0.7,
              "include_sources": true,
              "model_preference": "$modelPref"
            },
            "metadata": {
              "user_id": "$userId",
              "session_id": "gatling_session_${System.currentTimeMillis()}",
              "complexity": "$complexity",
              "timestamp": "${java.time.Instant.now()}"
            }
          }
          """
        }))
        .check(
          status.is(200),
          jsonPath("$.answer").exists,
          jsonPath("$.sources").exists,
          responseTimeInMillis.lte(5000),
          // Epic 8 SLO validations
          jsonPath("$.metrics.end_to_end_latency").optional.saveAs("endToEndLatency"),
          jsonPath("$.metrics.cost_usd").optional.saveAs("costUsd"),
          jsonPath("$.metrics.cache_hit").optional.saveAs("cacheHit"),
          jsonPath("$.metrics.api_gateway_latency").optional.saveAs("apiGatewayLatency"),
          jsonPath("$.metrics.generator_latency").optional.saveAs("generatorLatency"),
          jsonPath("$.metrics.analyzer_latency").optional.saveAs("analyzerLatency"),
          jsonPath("$.metrics.retriever_latency").optional.saveAs("retrieverLatency")
        )
    )
    .exec(session => {
      // Swiss engineering precision - log detailed metrics
      val endToEndLatency = session.attributes.get("endToEndLatency")
      val costUsd = session.attributes.get("costUsd")
      val cacheHit = session.attributes.get("cacheHit")

      if (endToEndLatency.isDefined) {
        println(s"End-to-end latency: ${endToEndLatency.get}ms")
      }
      if (costUsd.isDefined) {
        println(s"Cost per query: $${costUsd.get}")
      }
      if (cacheHit.isDefined) {
        println(s"Cache hit: ${cacheHit.get}")
      }

      session
    })
    .pause(2, 8) // Realistic think time

  // Health Check Scenario
  val healthCheckScenario = scenario("Health Check")
    .exec(
      http("Health Check")
        .get("/health")
        .check(
          status.is(200),
          responseTimeInMillis.lte(1000)
        )
    )
    .pause(30) // Check every 30 seconds

  // Auto-scaling Trigger Scenario
  val autoScalingTriggerScenario = scenario("Auto-scaling Trigger")
    .feed(queryFeeder)
    .feed(userIdFeeder)
    .exec(
      // Burst of requests to trigger HPA scaling
      repeat(10, "burstIndex") {
        exec(
          http("Burst Request ${burstIndex}")
            .post(s"/$apiVersion/query")
            .body(StringBody(session => {
              val (query, complexity) = session("queries").as[(String, String)]
              val userId = session("userId").as[String]

              s"""
              {
                "query": "$query",
                "options": {
                  "max_tokens": 500,
                  "model_preference": "fast"
                },
                "metadata": {
                  "user_id": "$userId",
                  "test_type": "autoscaling_trigger",
                  "burst_index": "${session("burstIndex").as[String]}",
                  "complexity": "$complexity"
                }
              }
              """
            }))
            .check(status.in(200, 429)) // Accept rate limiting
        ).pause(100.milliseconds) // Minimal pause between burst requests
      }
    )
    .pause(5) // Wait between burst groups

  // Cost Optimization Scenario
  val costOptimizationScenario = scenario("Cost Optimization")
    .feed(queryFeeder)
    .feed(userIdFeeder)
    .exec(
      http("Cost-Optimized Query")
        .post(s"/$apiVersion/query")
        .body(StringBody(session => {
          val (query, complexity) = session("queries").as[(String, String)]
          val userId = session("userId").as[String]

          s"""
          {
            "query": "$query",
            "options": {
              "max_tokens": 300,
              "model_preference": "fast",
              "cost_optimization": true
            },
            "metadata": {
              "user_id": "$userId",
              "test_type": "cost_optimization",
              "complexity": "$complexity"
            }
          }
          """
        }))
        .check(
          status.is(200),
          jsonPath("$.metrics.cost_usd").lte(0.01), // Cost SLO validation
          responseTimeInMillis.lte(2000) // Faster response for cost-optimized
        )
    )
    .pause(3, 6)

  // Load Testing Simulation Setup
  setUp(
    // 1. Baseline Load Test (Normal traffic)
    ragQueryScenario.inject(
      rampUsers(50).during(2.minutes),
      constantUsersPerSec(25).during(10.minutes),
      rampUsers(100).during(2.minutes),
      constantUsersPerSec(50).during(10.minutes),
      rampUsersPerSec(0).to(0).during(2.minutes)
    ).protocols(httpProtocol),

    // 2. Peak Load Test (Traffic spikes)
    ragQueryScenario.inject(
      nothingFor(15.minutes), // Start after baseline
      rampUsers(200).during(1.minute),
      rampUsers(500).during(2.minutes),
      constantUsersPerSec(250).during(5.minutes),
      rampUsersPerSec(250).to(100).during(2.minutes),
      rampUsersPerSec(100).to(0).during(1.minute)
    ).protocols(httpProtocol),

    // 3. Stress Test (Beyond normal capacity)
    ragQueryScenario.inject(
      nothingFor(25.minutes), // Start after peak load
      rampUsers(500).during(2.minutes),
      rampUsers(1000).during(3.minutes),
      constantUsersPerSec(500).during(5.minutes),
      rampUsersPerSec(500).to(250).during(2.minutes),
      rampUsersPerSec(250).to(0).during(2.minutes)
    ).protocols(httpProtocol),

    // 4. Spike Test (Sudden load increase)
    ragQueryScenario.inject(
      nothingFor(40.minutes), // Start after stress test
      constantUsersPerSec(50).during(30.seconds),
      rampUsersPerSec(50).to(750).during(30.seconds), // Sudden spike
      constantUsersPerSec(750).during(1.minute),
      rampUsersPerSec(750).to(50).during(30.seconds),
      constantUsersPerSec(50).during(30.seconds),
      rampUsersPerSec(50).to(0).during(30.seconds)
    ).protocols(httpProtocol),

    // 5. Endurance Test (Long duration)
    ragQueryScenario.inject(
      nothingFor(45.minutes), // Start after spike test
      rampUsers(200).during(2.minutes),
      constantUsersPerSec(100).during(30.minutes),
      rampUsersPerSec(100).to(0).during(2.minutes)
    ).protocols(httpProtocol),

    // 6. Auto-scaling Validation
    autoScalingTriggerScenario.inject(
      nothingFor(80.minutes), // Start after endurance
      constantUsersPerSec(10).during(2.minutes),
      rampUsersPerSec(10).to(50).during(2.minutes),
      constantUsersPerSec(50).during(5.minutes), // Should trigger HPA scaling
      rampUsersPerSec(50).to(10).during(2.minutes),
      constantUsersPerSec(10).during(2.minutes)
    ).protocols(httpProtocol),

    // 7. Cost Optimization Testing
    costOptimizationScenario.inject(
      nothingFor(95.minutes), // Start after autoscaling
      rampUsers(100).during(1.minute),
      constantUsersPerSec(50).during(10.minutes),
      rampUsersPerSec(50).to(0).during(1.minute)
    ).protocols(httpProtocol),

    // 8. Continuous Health Monitoring
    healthCheckScenario.inject(
      constantUsersPerSec(1).during(120.minutes) // Run throughout entire test
    ).protocols(httpProtocol)

  ).assertions(
    // Swiss Engineering SLO Assertions
    global.responseTime.percentile3.lte(2000), // P95 < 2s (Epic 8 SLO)
    global.responseTime.percentile4.lte(5000), // P99 < 5s
    global.successfulRequests.percent.gte(99.5), // Success rate > 99.5%
    global.responseTime.mean.lte(1000), // Mean response time < 1s

    // Service-specific assertions
    forAll.responseTime.percentile3.lte(5000), // All services P95 < 5s
    forAll.failedRequests.percent.lte(1), // Error rate < 1%

    // Resource efficiency assertions
    global.requestsPerSec.gte(10), // Minimum throughput

    // Cost optimization assertions (would need custom metrics)
    details("Cost-Optimized Query").responseTime.percentile3.lte(2000)
  ).protocols(httpProtocol)

  // Setup and teardown
  before {
    println("🚀 Starting Epic 8 Gatling Load Testing Suite")
    println(s"📍 Target: $baseUrl")
    println(s"🔑 API Version: $apiVersion")
    println(s"🌍 Environment: $environment")
    println("🇨🇭 Swiss Engineering Standards: Precision, Efficiency, Reliability")
  }

  after {
    println("🏁 Epic 8 Gatling Load Testing Complete")
    println("📊 Check reports for detailed Swiss engineering metrics")
  }
}